xarray/tests/test_coding_times.py

```diff
@@ -1886,45 +1889,56 @@ def test_decode_timedelta_via_units(
 
 
 _DECODE_TIMEDELTA_VIA_DTYPE_TESTS = {
-    "default": (True, None, np.dtype("timedelta64[ns]")),
-    "decode_timedelta=False": (True, False, np.dtype("int64")),
-    "decode_timedelta=True": (True, True, np.dtype("timedelta64[ns]")),
+    "default": (True, None, "ns", np.dtype("timedelta64[ns]")),
+    "decode_timedelta=False": (True, False, "ns", np.dtype("int64")),
+    "decode_timedelta=True": (True, True, "ns", np.dtype("timedelta64[ns]")),
+    "use-original-units": (True, True, "s", np.dtype("timedelta64[s]")),
     "inherit-time_unit-from-decode_times": (
         CFDatetimeCoder(time_unit="s"),
         None,
+        "ns",
         np.dtype("timedelta64[s]"),
     ),
     "set-time_unit-via-CFTimedeltaCoder-decode_times=True": (
         True,
         CFTimedeltaCoder(time_unit="s"),
+        "ns",
         np.dtype("timedelta64[s]"),
     ),
     "set-time_unit-via-CFTimedeltaCoder-decode_times=False": (
         False,
         CFTimedeltaCoder(time_unit="s"),
+        "ns",
         np.dtype("timedelta64[s]"),
     ),
     "override-time_unit-from-decode_times": (
         CFDatetimeCoder(time_unit="ns"),
         CFTimedeltaCoder(time_unit="s"),
+        "ns",
         np.dtype("timedelta64[s]"),
     ),
+    "decode-different-units": (
+        True,
+        CFTimedeltaCoder(time_unit="us"),
+        "s",
+        np.dtype("timedelta64[us]"),
+    ),
 }
 
 
 @pytest.mark.parametrize(
-    ("decode_times", "decode_timedelta", "expected_dtype"),
+    ("decode_times", "decode_timedelta", "original_unit", "expected_dtype"),
     list(_DECODE_TIMEDELTA_VIA_DTYPE_TESTS.values()),
     ids=list(_DECODE_TIMEDELTA_VIA_DTYPE_TESTS.keys()),
 )
 def test_decode_timedelta_via_dtype(
-    decode_times, decode_timedelta, expected_dtype
+    decode_times, decode_timedelta, original_unit, expected_dtype
 ) -> None:
-    timedeltas = pd.timedelta_range(0, freq="D", periods=3)
+    timedeltas = pd.timedelta_range(0, freq="D", periods=3, unit=original_unit)
```

Comment:
I do not fully understand why `mypy` complains here, but [we have worked around it this way in the past](https://github.com/pydata/xarray/blob/22bd964dca96226b2ed0b57762976743a9fe9e25/xarray/tests/test_coding_times.py#L1766):
```suggestion
    timedeltas = pd.timedelta_range(0, freq="D", periods=3, unit=original_unit)  # type: ignore[call-arg]
```

Reply:
I think this is due to the fact that the `unit` argument has not yet been added for [`timedelta_range` in pandas-stubs](https://github.com/pandas-dev/pandas-stubs/blob/5b38eb5d20ff00216450c4fefd0ec2781497fd12/pandas-stubs/core/indexes/timedeltas.pyi#L79-L86).  I went ahead and made a PR: https://github.com/pandas-dev/pandas-stubs/pull/1271.